src/APIBasedAuth.ts

```diff
@@ -0,0 +1,192 @@
+
+    this.clientOptions = options;
+  }
+
+  /* istanbul ignore next */
+  private async _setAuthorizationHeader(
+    request: AxiosRequestConfig<unknown>
+  ): Promise<AxiosRequestConfig<unknown>> {
+    if (this.tokenData?.access_token) {
+      request.headers = {
+        ...request?.headers,
+        Authorization: `Bearer ${this.tokenData.access_token}`
+      };
+    }
+    return request;
+  }
+
+  public confirmPasswordlessAuth(
+    baseInput: Omit<APIBasedAuth.VerifyPasswordlessAuthData, 'clientId'>
+  ) {
+    const input = {
+      ...baseInput,
+      // if session or username is not present get stored properties
+      ...(!baseInput.session || !baseInput.username ? this.session : null)
+    };
+    return this.client
+      .post<
+        APIBasedAuth.SuccessfulAuthResponse,
+        AxiosResponse<APIBasedAuth.SuccessfulAuthResponse>,
+        APIBasedAuth.VerifyPasswordlessAuthData
+      >('/passwordless-auth/verify', {
+        clientId: this.clientOptions.clientId,
+        code: input.code,
+        session: input.session,
+        username: input.username
+      })
+      .then(({ data }) => {
+        this.tokenData = {
+          access_token: data.accessToken,
+          id_token: data.idToken,
+          refresh_token: data.refreshToken
+        };
+        // TODO: add storage
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  public initiatePasswordAuth(
+    input: Omit<APIBasedAuth.SignInData, 'clientId'>
+  ) {
+    return this.client
+      .post<
+        APIBasedAuth.SuccessfulAuthResponse,
+        AxiosResponse<APIBasedAuth.SuccessfulAuthResponse>,
+        APIBasedAuth.SignInData
+      >('/login', {
+        clientId: this.clientOptions.clientId,
+        ...input
+      })
+      .then(({ data }) => {
+        this.tokenData = {
+          access_token: data.accessToken,
+          id_token: data.idToken,
+          refresh_token: data.refreshToken
+        };
+        // TODO: add storage
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  public initiatePasswordlessAuth({
+    appsBaseUri,
+    loginAppBasePath,
+    username
+  }: Omit<APIBasedAuth.PasswordlessAuthData, 'clientId'>) {
+    return this.client
+      .post<
+        APIBasedAuth.PasswordlessAuthResponse,
+        AxiosResponse<APIBasedAuth.PasswordlessAuthResponse>,
+        APIBasedAuth.PasswordlessAuthData
+      >('/passwordless-auth', {
+        appsBaseUri,
+        clientId: this.clientOptions.clientId,
+        loginAppBasePath,
+        username
+      })
+      .then(({ data }) => {
+        this.session = {
+          session: data.session,
+          username
+        };
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  /* istanbul ignore next */
+  public logout() {
+    this.session = undefined;
+    this.tokenData = undefined;
+  }
+}
+
+/* istanbul ignore next */
+declare namespace APIBasedAuth {
```

Comment:
@smolster this sort of stuff is what you had in mind for using https://github.com/lifeomic/one-schema/ , right?